File: src/cli/auth/workspace.ts
import type { Command } from "commander";
import {
  getWorkspaces,
  getDefaultWorkspace,
  setDefaultWorkspace,
  removeWorkspace,
} from "../../lib/config.ts";
import { printError, printJson } from "../../lib/output.ts";

export function registerWorkspace(auth: Command): void {
  const workspace = auth.command("workspace").description("Manage workspace profiles");

  workspace
    .command("list")
    .description("List all stored workspaces")
    .action(() => {
      const workspaces = getWorkspaces();
      const defaultWs = getDefaultWorkspace();
      const items = Object.entries(workspaces).map(([alias, ws]) => ({
        alias,
        name: ws.name,
        urlKey: ws.urlKey,
        default: alias === defaultWs,
      }));
      printJson({ items });
    });

  workspace
    .command("switch")
    .description("Set default workspace")
    .argument("<alias>", "Workspace alias to switch to")
    .action((alias: string) => {
      try {
        setDefaultWorkspace(alias);
        printJson({ ok: true, default_workspace: alias });
      } catch (err) {
        printError(err instanceof Error ? err.message : "Switch failed");
      }
    });

  workspace
    .command("remove")
    .description("Remove a stored workspace")
    .argument("<alias>", "Workspace alias to remove")
    .action((alias: string) => {
      try {
        const wasDefault = getDefaultWorkspace() === alias;
        removeWorkspace(alias);
        const newDefault = getDefaultWorkspace();
        printJson({
          ok: true,
          removed: alias,
          ...(wasDefault && { warning: "Removed the default workspace" }),
          default_workspace: newDefault ?? null,
        });
      } catch (err) {
        printError(err instanceof Error ? err.message : "Remove failed");
      }
    });
}

File: internal/cli/auth/workspace.go
package auth

import (
	"github.com/spf13/cobra"

	"github.com/shhac/lin/internal/config"
	"github.com/shhac/lin/internal/output"
)

func registerWorkspace(auth *cobra.Command) {
	workspace := &cobra.Command{
		Use:   "workspace",
		Short: "Manage workspace profiles",
	}
	output.HandleUnknownCommand(workspace, "Run 'lin auth workspace list' for available workspaces")

	registerWorkspaceList(workspace)
	registerWorkspaceSwitch(workspace)
	registerWorkspaceRemove(workspace)

	auth.AddCommand(workspace)
}

func registerWorkspaceList(workspace *cobra.Command) {
	cmd := &cobra.Command{
		Use:   "list",
		Short: "List all stored workspaces",
		Args:  cobra.NoArgs,
		Run: func(_ *cobra.Command, _ []string) {
			workspaces := config.GetWorkspaces()
			defaultWs := config.GetDefaultWorkspace()
			items := make([]map[string]any, 0, len(workspaces))
			for alias, ws := range workspaces {
				items = append(items, map[string]any{
					"alias":   alias,
					"name":    ws.Name,
					"urlKey":  ws.URLKey,
					"default": alias == defaultWs,
				})
			}
			output.PrintJSON(map[string]any{"items": items})
		},
	}
	workspace.AddCommand(cmd)
}

func registerWorkspaceSwitch(workspace *cobra.Command) {
	cmd := &cobra.Command{
		Use:   "switch <alias>",
		Short: "Set default workspace",
		Args:  cobra.ExactArgs(1),
		Run: func(_ *cobra.Command, args []string) {
			alias := args[0]
			if err := config.SetDefaultWorkspace(alias); err != nil {
				output.PrintError(err.Error())
				return
			}
			output.PrintJSON(map[string]any{"ok": true, "default_workspace": alias})
		},
	}
	workspace.AddCommand(cmd)
}

func registerWorkspaceRemove(workspace *cobra.Command) {
	cmd := &cobra.Command{
		Use:   "remove <alias>",
		Short: "Remove a stored workspace",
		Args:  cobra.ExactArgs(1),
		Run: func(_ *cobra.Command, args []string) {
			alias := args[0]
			wasDefault := config.GetDefaultWorkspace() == alias
			if err := config.RemoveWorkspace(alias); err != nil {
				output.PrintError(err.Error())
				return
			}
			newDefault := config.GetDefaultWorkspace()
			result := map[string]any{
				"ok":                true,
				"removed":           alias,
				"default_workspace": nilIfEmpty(newDefault),
			}
			if wasDefault {
				result["warning"] = "Removed the default workspace"
			}
			output.PrintJSON(result)
		},
	}
	workspace.AddCommand(cmd)
}
